<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD>
<META http-equiv=Content-Type content="text/html; charset=windows-1252">
<META content="MSHTML 6.00.2800.1226" name=GENERATOR></HEAD>
<BODY>
<TABLE>
  <TBODY>
  <tr>
   <td><b>Object</b></td><td><b>Description</b></td>
  </tr>
  <TR>
    <TD vAlign=top>Agent</TD>
    <TD>A therapeutic agent (drug, intervention therapy) used in a clinical 
      trial protocol. Application: used primarily by&nbsp;CMAP and&nbsp;EVS 
      applications. Related domain objects: ClinicalTrialProtocol, Target&nbsp; 
    </TD></TR>
  <TR>
    <TD vAlign=top>Anomaly</TD>
    <TD>An irregularity in either the expression of a gene&nbsp;or its 
      structure (i.e., a mutation). Application: defined and used by the CMAP 
      project. Related domain objects: Histopathology, Target </TD></TR>
  <TR>
    <TD vAlign=top>Chromosome</TD>
    <TD>An object representing a specific chromosome for a specific taxon; 
      provides access to all known genes contained in the chromosome and to the 
      taxon. Application: used by CMAP and other applications to reason about 
      the molecular basis of cancer. Related domain objects: Gene, Taxon </TD></TR>
  <TR>
    <TD vAlign=top>ClinicalTrialProtocol</TD>
    <TD>The protocol associated with a clinical trial; organizes 
      administrative information about the trial such as Organization ID, 
      participants, phase, etc., and provides access to the administered Agents. 
      Application: used primarily by&nbsp;CMAP. Related domain objects: Agent, 
      ProtocolAssociation </TD></TR>
  <TR>
    <TD vAlign=top>Clone</TD>
    <TD>An object used to hold information pertaining to I.M.A.G.E. clones; 
      provides access to sequence information, associated trace files, and the 
      clone’s library. Application: The caBIO Clone data are imported from the 
      CGAP website databases. Related domain objects: Sequence, Library, 
      TraceFile, SNP </TD></TR>
  <TR>
    <TD vAlign=top>CMAPOntology</TD>
    <TD>An object providing entry to the CMAP gene ontology, which categorizes 
      genes by function; provides access to gene objects corresponding to the 
      ontological term, as well as to ancestor and descendant terms in the 
      ontology tree. Application: defined and used by&nbsp;CMAP applications. 
      Related domain objects: CMAPOntologyRelationship, Gene </TD></TR>
  <TR>
    <TD vAlign=top>CMAPOntologyRelationship</TD>
    <TD>An object specifying a child or parent relationship between 
      CMAPOntology objects. Application: used and defined by&nbsp;CMAP 
      applications. Related domain objects: CMAPOntology </TD></TR>
  <TR>
    <TD vAlign=top>Disease</TD>
    <TD>Disease objects specify a disease name and ID; disease objects also 
      provide access to: ontological relations to other diseases; clinical trial 
      protocols treating the disease; and specific histologies associated with 
      instances of the disease. Application: used by the CMAP project. Related 
      domain objects: ClinicalTrialProtocol, Histopathology, DiseaseRelationship 
      Extends java.lang.Object Implements: XMLInterface, java.io.Serializable, 
      Ontologable </TD></TR>
  <TR>
    <TD vAlign=top>DiseaseRelationship</TD>
    <TD>An object specifying a child or parent relationship between Disease 
      objects. Application: used by the CMAP project. Related domain objects: 
      Disease </TD></TR>
  <TR>
    <TD vAlign=top>ExpressionFeature</TD>
    <TD>Associated with a Gene object through the gene’s method 
      getExpressionFeature(); provides access to the list of organs where the 
      gene is known to be expressed. </TD></TR>
  <TR>
    <TD vAlign=top>ExpressionMeasurement</TD>
    <TD>An object representing a structure that is capable of measuring the 
      absolute or relative amount of a given compound. Related domain objects: 
      Gene, Sequence </TD></TR>
  <TR>
    <TD vAlign=top>ExpressionMeasurementArray</TD>
    <TD>An array of ExpressionMeasurement objects. Related domain objects: 
      Gene, Sequence </TD></TR>
  <TR>
    <TD vAlign=top>Gene</TD>
    <TD>Gene objects are the effective portal to most of the genomic 
      information provided by the caBIO data services; organs, diseases, 
      chromosomes, pathways, sequence data, and expression experiments are among 
      the many objects accessible via a gene. Related domain objects: 
      ExpressionFeature, Organ, Disease, Chromosome, Taxon, Sequence, GeneAlias, 
      GeneHomolog, MapLocation, Protein, SNP, Target, ExpressionMeasurement, 
      Pathway, GoOntology </TD></TR>
  <TR>
    <TD vAlign=top>GeneAlias</TD>
    <TD>An alternative name for a gene; provides descriptive information about 
      the gene (as it is known by this alias), as well as access to the Gene 
      object it refers to. Related Domain Objects: Gene </TD></TR>
  <TR>
    <TD vAlign=top>GeneHomolog</TD>
    <TD>Defined only in relation to another Gene object of interest, the 
      GeneHomolog in caBIO is the functional equivalent of that gene in another 
      taxon (i.e., its ortholog). The GeneHomolog is a specialization of the 
      parent Gene object; in addition to all of the methods provided by the gene 
      interface, the homolog object provides the percent of sequence similarity 
      of the homolog to the related gene of interest. Related domain objects: 
      Gene, ExpressionFeature, Organ, Disease, Chromosome, Taxon, Sequence, 
      GeneAlias, GeneHomolog, MapLocation, Protein, SNP, Target, 
      ExpressionMeasurement, Pathway, GoOntology </TD></TR>
  <TR>
    <TD vAlign=top>GoOntology</TD>
    <TD>An object providing entry to a Gene object’s position in the Gene 
      Ontology Consortium’s controlled vocabularies, as recorded by LocusLink; 
      provides access to gene objects corresponding to the ontological term, as 
      well as to ancestor and descendant terms in the ontology tree. Related 
      domain objects: Gene, GoOntologyRelationship </TD></TR>
  <TR>
    <TD vAlign=top>GoOntologyRelationship</TD>
    <TD>An object specifying a child or parent relationship between GoOntology 
      objects. Related domain objects: GoOntology </TD></TR>
  <TR>
    <TD vAlign=top>Histopathology</TD>
    <TD>An object representing anatomical changes in a diseased tissue sample 
      associated with an expression experiment; captures the relationship 
      between organ and disease. Application: used by the CMAP project. Related 
      domain objects: Anomaly, Organ, Disease, ExpressionExperiment, 
      ESTExperiment, SAGEExperiment </TD></TR>
  <TR>
    <TD vAlign=top>Library</TD>
    <TD>An object representing a CGAP library; provides access to information 
      about: the tissue sample and its method of preparation, the library 
      protocol that was used, the clones contained in the library, and the 
      sequence information derived from the library. Application: The caBIO 
      libraries are extracted from the CGAP databases. Related domain objects: 
      Clone, Sequence, Tissue, Protocol </TD></TR>
  <TR>
    <TD vAlign=top>MapLocation</TD>
    <TD>Associated with a Gene object, the physical map location of the gene. 
      Related domain objects: Chromosome, Gene, Taxon. </TD></TR>
  <TR>
    <TD vAlign=top>Organ</TD>
    <TD>A representation of an organ whose name occurs in a controlled 
      vocabulary; provides access to any Histopathology objects for the organ, 
      and, because it is “ontolog-able,” provides access to its ancestral and 
      descendant terms in the vocabulary. Related domain objects: 
      Histopathology, OrganRelationship </TD></TR>
  <TR>
    <TD vAlign=top>OrganRelationship</TD>
    <TD>An object specifying a child or parent relationship between Organ 
      objects. Related domain objects: Organ </TD></TR>
  <TR>
    <TD vAlign=top>Pathway</TD>
    <TD>An object representation of a molecular/cellular pathway compiled by 
      BioCarta. Pathways are associated with specific Taxon objects, and contain 
      multiple Gene objects, which may be Targets for treatment. Related domain 
      objects: Gene, Taxon, Target. </TD></TR>
  <TR>
    <TD vAlign=top>Protein</TD>
    <TD>An object representation of a protein; provides access to the encoding 
      gene via its GenBank ID, the taxon in which this instance of the protein 
      occurs, and references to homologous proteins in other species. Related 
      domain objects: Gene, ProteinHomolog, Taxon. </TD></TR>
  <TR>
    <TD vAlign=top>ProteinHomolog</TD>
    <TD>Defined only in relation to another Protein object of interest, the 
      ProteinHomolog in caBIO is the functional equivalent of that protein in 
      another taxon (i.e., its ortholog). The ProteinHomolog is a specialization 
      of the parent Protein object; in addition to the methods provided by the 
      protein interface, the Homolog object provides the percent of sequence 
      similarity of the homolog to the related protein of interest. Related 
      domain objects: Gene, Protein, Taxon. </TD></TR>
  <TR>
    <TD vAlign=top>Protocol</TD>
    <TD>An object representation of the protocol used in assembling a clone 
      library. </TD></TR>
  <TR>
    <TD vAlign=top>ProtocolAssociation</TD>
    <TD>An association class relating between ClinicalTrialProtocols to 
      Diseases. Application: used primarily by the CMAP project. Related domain 
      objects: ConceptSearch, ClinicalTrialProtocol, Disease </TD></TR>
  <TR>
    <TD vAlign=top>Sequence</TD>
    <TD>An object representation of a gene sequence; provides access to the 
      clones from which it was derived, the ASCII representation of the residues 
      it contains, and the sequence ID. Related domain objects: Clone, Gene, 
      Protein </TD></TR>
  <TR>
    <TD vAlign=top>SNP</TD>
    <TD>An object representing a Single Nucleotide Polymorphism; provides 
      access to the clones and the trace files from which it was identified, the 
      two most common substitutions at that position, the offset of the SNP in 
      the parent sequence, and a confidence score. Application: The SNPs 
      provided by caBIO were identified by the GAI project. Related domain 
      objects: Clone, TraceFile </TD></TR>
  <TR>
    <TD vAlign=top>Target</TD>
    <TD>A gene thought to be at the root of a disease etiology, and which is 
      targeted for therapeutic intervention in a clinical trial. Application: 
      defined and used by the CMAP project. Related domain objects: Agent, 
      Anomaly, Gene </TD></TR>
  <TR>
    <TD vAlign=top>Taxon</TD>
    <TD>An object representing the various names (scientific, common, 
      abbreviated, etc.) for a species associated with a specific instance of a 
      Gene, Chromosome, Pathway, Protein, or Tissue. Related domain objects: 
      Gene, Chromosome, Pathway, Protein, Tissue </TD></TR>
  <TR>
    <TD vAlign=top>Tissue</TD>
    <TD>A group of similar cells united to perform a specific function. 
      Related domain objects: Disease, Organ, Protocol, Taxon </TD></TR>
  <TR>
    <TD vAlign=top>ClassSchemeClassSchemeItem</TD>
    <TD>This component is used to associate a set of classification scheme 
      items with a particular classification scheme, and to store details of 
      that association such as the display order of the items within that 
      scheme. Application: defined and used by&nbsp;the caDSR project for 
      implementation purposes. Related domain objects: ClassificationScheme, 
      ClassificationSchemeItem. </TD></TR>
  <TR>
    <TD vAlign=top>ValueMeaning</TD>
    <TD>The significance associated with an allowable/permissible value. 
      Application: A component of the ISO/IEC 11179 model, implemented in caDSR. 
      Related domain objects: ConceptualDomain, PermissibleValue </TD></TR>
  <TR>
    <TD vAlign=top>AnimalModel</TD>
    <TD>A strain of animals used to study the various types of cancer. Related 
      MMHCC domain objects: Availability, CarcinogenicIntervention, CellLine, 
      GenomicSegment, Histopathology, Image, InducedMutation, JaxInfo, 
      MicroArrayData, PartyRole, Phenotype, Publication, RepositoryInfo, 
      TargetedModification, Therapy, Transgene, Xenograft Other related domain 
      objects: gov.nih.nci.common.util.Credential, 
      gov.nih.nci.common.util.ApprovalStatus, gov.nih.nci.cabio.bean.taxon 
</TD></TR>
  <TR>
    <TD vAlign=top>Availability</TD>
    <TD>The availability status of a developed strain from the MMHCC 
      Repository, the Jackson Laboratory, or directly from the principal 
      investigator. </TD></TR>
  <TR>
    <TD vAlign=top>CarcinogenicIntervention</TD>
    <TD>A treatment (chemical or drug administration, radiation, etc) applied 
      to an animal model to induce a disease state. Related MMHCC domain 
      objects: EnvironmentalFactor, GeneDelivery, TreatmentSchedule Other 
      related domain objects: gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>CellLine</TD>
    <TD>A cell line generated from a particular strain of animal model. 
      Related MMHCC domain objects: Organ, Publication Other related domain 
      objects: gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>Conditionality</TD>
    <TD>Indicates if a transgene or targeted modification is time- or 
      tissue-specific. </TD></TR>
  <TR>
    <TD vAlign=top>ContactInfo</TD>
    <TD>Contact information for the person who submitted data for the selected 
      model. Related MMHCC domain objects: Party </TD></TR>
  <TR>
    <TD vAlign=top>Disease</TD>
    <TD>A disease state recognized and identified in a controlled vocabulary 
      maintained by the EVS project. Related MMHCC domain objects: 
      Histopathology </TD></TR>
  <TR>
    <TD vAlign=top>EngineeredGene</TD>
    <TD>A gene sequence that has been genetically modified to induce a desired 
      state in an animal model. Related MMHCC domain objects: Conditionality, 
      GeneFunction, GenotypeSummary, Image, Promoter Other related domain 
      objects: gov.nih.nci.common.util.Credential, 
      gov.nci.nih.gov.cabio.bean.ExpressionFeature </TD></TR>
  <TR>
    <TD vAlign=top>EnvironmentalFactor</TD>
    <TD>A chemical, radiation, or hormone treatment, or other environmental 
      factor that initiates or supports development of neoplasis in the animal 
      model. </TD></TR>
  <TR>
    <TD vAlign=top>GeneDelivery</TD>
    <TD>The method of introducing a modified gene to the recipient animal. 
      Related MMHCC domain objects: EngineeredGene, Organ Other related domain 
      objects: gov.nci.nih.gov.cabio.bean.Gene </TD></TR>
  <TR>
    <TD vAlign=top>GeneFunction</TD>
    <TD>The known or hypothesized function of a gene. </TD></TR>
  <TR>
    <TD vAlign=top>GeneticAlteration</TD>
    <TD>Genetic alterations found in the diseased tissue other than the 
      engineered genetic changes. </TD></TR>
  <TR>
    <TD vAlign=top>GenomicSegment</TD>
    <TD>A region or set of regions of a genome including chromosome, gene, 
      breakpoint etc. The size of a genomic segment varies from a fraction of a 
      gene to a region containing many genes surrounded by non-coding sequences, 
      and can be as large as a chromosome. Related MMHCC domain objects: 
      EngineeredGene, IntegrationType, SegmentType Other related domain objects: 
      gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>GenotypeSummary</TD>
    <TD>Summary of genotype information for a particular model Related MMHCC 
      domain objects: Nomenclature </TD></TR>
  <TR>
    <TD vAlign=top>Histopathology</TD>
    <TD>Microscopic structure of diseased tissues/organs from a model animal. 
      Related MMHCC domain objects: Disease, GeneticAlteration, Histopathology, 
      Organ Other related domain objects: gov.nih.nci.common.util.Credential 
  </TD></TR>
  <TR>
    <TD vAlign=top>Image</TD>
    <TD>Images and image annotations associated with animal models. Related 
      MMHCC domain objects: Availability Other related domain objects: 
      gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>InducedMutation</TD>
    <TD>A mutation in a gene caused by exposure to chemicals, radiation or 
      other types of mutagens. By definition, an induced mutation is inherited 
      by the next generation. Related MMHCC domain objects: EnvironmentalFactor, 
      EngineeredGene Other related domain objects: 
      gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>IntegrationType</TD>
    <TD>Location of the integration of the engineered gene, e.g. “random” or 
      “targeted”. </TD></TR>
  <TR>
    <TD vAlign=top>JaxInfo</TD>
    <TD>Information associated with an animal model in The Jackson Laboratory; 
      for example: stock number, strain name, etc. </TD></TR>
  <TR>
    <TD vAlign=top>MicroArrayData</TD>
    <TD>Gene expression data from microarray experiments using cells or 
      tissues from an animal model. Related MMHCC domain objects: Availability 
      Other related domain objects: gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>Nomenclature</TD>
    <TD></TD></TR>
  <TR>
    <TD vAlign=top>Organ</TD>
    <TD>An organ from a model animal, as recognized and identified in a 
      controlled vocabulary maintained by the EVS project. </TD></TR>
  <TR>
    <TD vAlign=top>Organization</TD>
    <TD>The institution (university, institute, laboratory) where a 
      contributing scientist works. </TD></TR>
  <TR>
    <TD vAlign=top>Party</TD>
    <TD>A Person or group of persons (Organization) having a designated 
      relationship (PartyRole) to parts of the MMHCC data with explicit access 
      permissions. Related MMHCC domain objects: ContactInfo, PartyRole </TD></TR>
  <TR>
    <TD vAlign=top>PartyRole</TD>
    <TD>The PartyRole is used to map designated MMHCC users to the read/write 
      access privileges associated with MMHCC data. Related MMHCC domain 
      objects: AnimalModel </TD></TR>
  <TR>
    <TD vAlign=top>Person</TD>
    <TD>An individual involved in the deposition, review, and/or approval of 
      data in MMHCC. </TD></TR>
  <TR>
    <TD vAlign=top>Phenotype</TD>
    <TD>The physical appearance or otherwise observable characteristics of a 
      model animal. Related MMHCC domain objects: SexDistribution Other related 
      domain objects: gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>Promoter</TD>
    <TD>A region of DNA sequence upstream of the coding region to which RNA 
      polymerase will bind and initiate replication. Related MMHCC domain 
      objects: RegulatoryElementType Other related domain objects: 
      gov.nih.nci.common.util.Credential, gov.nih.nci.cabio.bean.Taxon </TD></TR>
  <TR>
    <TD vAlign=top>Publication</TD>
    <TD>Publications describing the animal model itself or experiments in 
      which the animal model was used Related MMHCC domain objects: 
      PublicationStatus </TD></TR>
  <TR>
    <TD vAlign=top>PublicationStatus</TD>
    <TD>Status regarding a scientific paper, e.g., “unpublished”, “submitted”, 
      “published”. </TD></TR>
  <TR>
    <TD vAlign=top>RegulatoryElement</TD>
    <TD>A region of DNA sequence controlling the transcription/expression of a 
      gene. Related MMHCC domain objects: RegulatoryElementType Other related 
      domain objects: gov.nih.nci.common.util.Credential, 
      gov.nih.nci.cabio.bean.Taxon </TD></TR>
  <TR>
    <TD vAlign=top>RegulatoryElementType</TD>
    <TD>The type of regulation imposed by the element, e.g., suppressor, 
      promoter, etc. </TD></TR>
  <TR>
    <TD vAlign=top>RepositoryInfo</TD>
    <TD>A RepositoryInfo object contains information about the availability of 
      a particular model from the MMHCC repository. Data submitters to the 
      cancer models database can indicate that their model should be submitted 
      to the repository for acceptance for acceptance. </TD></TR>
  <TR>
    <TD vAlign=top>Role</TD>
    <TD>Role that a person or organization plays; for example, public user, 
      submitter, reviewer, etc. Related MMHCC domain objects: PartyRole </TD></TR>
  <TR>
    <TD vAlign=top>SegmentType</TD>
    <TD>Genetic segment type such as chromosome, contig, CpG islands, 
      repetitive DNA (e.g. Alu, LINE, SINE etc.), gene, breakpoint etc. (see 
      GenomicSegment). </TD></TR>
  <TR>
    <TD vAlign=top>SexDistribution</TD>
    <TD>The observerable distribution of phenotypes between sexes. </TD></TR>
  <TR>
    <TD vAlign=top>TargetedModification</TD>
    <TD>Modification of a specific gene (versus one randomly selected) by a 
      genetic engineering technology called gene targeting through homologous 
      recombination; usually achieved using gene targeting vectors. Related 
      MMHCC domain objects: EngineeredGene, ModificationType Other related 
      domain objects: gov.nih.nci.common.util.Credential </TD></TR>
  <TR>
    <TD vAlign=top>Therapy</TD>
    <TD>A defined treatment protocol for testing the efficacy of the treatment 
      on an engineered animal model. Related MMHCC domain objects: Publication, 
      TreatmentSchedule, Other related domain objects: 
      gov.nih.nci.common.util.Credential, gov.nih.nci.cabio.bean.Agent </TD></TR>
  <TR>
    <TD vAlign=top>Transgene</TD>
    <TD>A foreign gene that has been integrated into the genome of an animal. 
      Related MMHCC domain objects: IntegrationType, RegulatoryElement Other 
      related domain objects: gov.nih.nci.common.util.Credential, 
      gov.nih.nci.cabio.bean.Taxon </TD></TR>
  <TR>
    <TD vAlign=top>TreatmentSchedule</TD>
    <TD>The dosage and regimen for treating cancer in an animal model. </TD></TR>
  <TR>
    <TD vAlign=top>Xenograft</TD>
    <TD>A surgical graft of tissue from one species onto or into individuals 
      of unlike species, genus or family. Related MMHCC domain objects: Organ 
      Other related domain objects: gov.nih.nci.common.util.Credential, 
      gov.nih.nci.cabio.bean.Taxon </TD></TR></TBODY></TABLE></BODY></HTML>
